### PR TITLE
Sipyeon

### PR DESCRIPTION
**Pull request by sypyeon**

### CODE REVIEW

File: src/server/Client.cpp [hunk 1]
```diff
@@ -1,5 +1,152 @@
...
+
+    if (req->remnants.compare(0, 3, "0\r\n") == 0) {
+      chunk_end = 0;
+    } else {
+      const size_t crlf_zero = req->remnants.find("\r\n0\r\n");
+      chunk_end =
+          (crlf_zero == std::string::npos) ? std::string::npos : crlf_zero + 2;
+    }
+    if (chunk_end != std::string::npos) {
+      const Result<size_t> unchunked = req->unchunk(chunk_end);
```
Comment (DavidLee18) [inline src/server/Client.cpp]: `TRY` 로 대체하면 좋을 것 같아요~

File: src/server/Client.cpp [hunk 2]
```diff
@@ -115,86 +262,24 @@ Result<Request *> Request::from_buff(std::string &buff) {
...
-      return ERR(Request *, Errors::bad_request);
-
-    if (cl_pos == std::string::npos && te_pos == std::string::npos &&
-        (method == POST || method == PUT || method == PATCH))
-      return ERR(Request *, Errors::malformed_header);
-  } else {
-    return ERR(Request *, Errors::internal_server_error);
+  // 1. Parse request line
+  {
+    Result<Void> _prl = Request::parse_request_line(
```
Comment (DavidLee18) [inline src/server/Client.cpp]: `TRY` 로 대체 2트

File: src/server/Client.cpp [hunk 3]
```diff
@@ -115,86 +262,24 @@ Result<Request *> Request::from_buff(std::string &buff) {
...
-    if (!line.empty() && line[line.size() - 1] == '\r')
-      line.erase(line.size() - 1);
-
-    const size_t colon_pos = line.find(':');
-    if (line.length() > 0x2000 || line[0] == ' ' || line[0] == '\t' ||
-        colon_pos == std::string::npos || colon_pos == 0 ||
-        line[colon_pos - 1] == ' ' || line[colon_pos - 1] == '\t') {
+  // Parse header lines into req->header
+  {
+    Result<Void> _rh = Request::parse_headers(ss, req);
```
Comment (DavidLee18) [inline src/server/Client.cpp]: `TRY` 3트

File: src/server/Client.cpp [hunk 4]
```diff
@@ -209,69 +294,12 @@ Result<Request *> Request::from_buff(std::string &buff) {
...
-             it != first_chunk_size.end(); ++it)
-          if (!((*it >= '0' && *it <= '9') || (*it >= 'a' && *it <= 'f') ||
-                (*it >= 'A' && *it <= 'F'))) {
-            delete req;
-            return ERR(Request *, Errors::bad_request);
-          }
-      }
-    }
-    return OK(Request *, req);
+  Result<Void> _pb = Request::parse_body_or_chunked(ss, buff, req, header_end);
```
Comment (DavidLee18) [inline src/server/Client.cpp]: `TRYF` 로 대체 바랍니다~

File: src/server/ResponseHandlers.cpp [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+
+Result<Void> ResponseHandlers::register_cgi(
+    const Request &request, const RouteRule_CGI &rule, EPoll *epoll,
+    std::map<std::string, std::string> const &cgi_interpreters,
+    std::map<FileDescriptor const *,
+             std::pair<FileDescriptor const *, CgiDelegate *> > &cgis,
+    FileDescriptor const *client_fd, char **envp) {
+  CgiDelegate *del =
+      static_cast<CgiDelegate *>(operator new(sizeof(CgiDelegate)));
+  const Result<CgiDelegate> del_ =
```
Comment (DavidLee18) [inline src/server/ResponseHandlers.cpp:26]: `TRYF` 대체 2트

File: src/server/ResponseHandlers.cpp [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+        }
+
+        std::string file_path = upload_path + "/";
+        file_path += filename;
+        std::cout << utils::info << "Uploading file: " << file_path
+                  << " (size: " << part_data.length() << ")" << std::endl;
+        std::cout << utils::info << "First 20 bytes (hex): ";
+        for (size_t i = 0;
+             i < std::min(static_cast<size_t>(20), part_data.length()); i++) {
+          printf("%02x ", static_cast<unsigned char>(part_data[i]));
```
Comment (DavidLee18) [inline src/server/ResponseHandlers.cpp]: `std::cout` 으로 수정해 주세요~

File: src/server/Server.cpp [hunk 7]
```diff
@@ -467,59 +106,84 @@ Result<Void> Server::start(char **envp) {
...
+          clients.find(fd);
+      if (jt == clients.end())
+        continue;
+      std::ostringstream oss;
+      oss << resp;
+      if (!resp.keep_alive)
+        jt->second.dropping = true;
+      jt->second.out_buff = oss.str();
+      {
+        Result<Void> cw = client_write(jt->first);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_` 로 대체 가능

File: src/server/Server.cpp [hunk 8]
```diff
@@ -467,59 +106,84 @@ Result<Void> Server::start(char **envp) {
...
+        if (!cw.has_value())
+          return ERR(Void, cw.error());
+      }
+      if (clients.find(fd) != clients.end() && jt->second.dropping) {
+        Result<Void> d = disconnect(fd);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 9]
```diff
@@ -467,59 +106,84 @@ Result<Void> Server::start(char **envp) {
...
+        if (!d.has_value())
+          return ERR(Void, d.error());
+      }
+    }
+
+    for (size_t i = 0; i < clients_to_disconnect.size(); ++i) {
+      Result<Void> d = disconnect(clients_to_disconnect[i]);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 10]
```diff
@@ -538,21 +202,32 @@ Result<Void> Server::start(char **envp) {
           if (!resp.keep_alive)
             jt->second.dropping = true;
           jt->second.out_buff = oss.str();
-          client_write(jt->first);
+          {
+            Result<Void> cw = client_write(jt->first);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 11]
```diff
@@ -569,17 +244,19 @@ Result<Void> Server::start(char **envp) {
...
         if (epoll_timeout == -1 ||
             cgi_remaining < static_cast<size_t>(epoll_timeout))
           epoll_timeout = static_cast<long>(cgi_remaining);
       }
     }
     for (std::set<CgiDelegate *>::const_iterator it = cgis_to_reap.begin();
-         it != cgis_to_reap.end(); ++it)
-      reap_cgi(*it);
+         it != cgis_to_reap.end(); ++it) {
+      Result<Void> rr = reap_cgi(*it);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 12]
```diff
@@ -596,25 +273,37 @@ Result<Void> Server::start(char **envp) {
...
       if (listeners.find(fd) != listeners.end()) {
-        new_connection(fd);
-      } else if (clients.find(fd) !=
-                 clients.end()) { // 2. 이미 연결된 클라이언트 소켓인 경우
-        if (event->in)
-          client_read(fd, envp);
-        if (event->out)
-          client_write(fd);
+        {
+          Result<Void> nc = new_connection(fd);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 13]
```diff
@@ -596,25 +273,37 @@ Result<Void> Server::start(char **envp) {
...
+          if (!nc.has_value())
+            return ERR(Void, nc.error());
+        }
+      } else if (clients.find(fd) != clients.end()) {
+        if (event->in) {
+          Result<Void> cr = client_read(fd, envp);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 14]
```diff
@@ -596,25 +273,37 @@ Result<Void> Server::start(char **envp) {
...
+          if (!cr.has_value())
+            return ERR(Void, cr.error());
+        }
+        if (event->out) {
+          Result<Void> cw = client_write(fd);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 15]
```diff
@@ -596,25 +273,37 @@ Result<Void> Server::start(char **envp) {
...
+          if (!cw.has_value())
+            return ERR(Void, cw.error());
+        }
         if (event->err || event->hup || event->rdhup) {
-          if (clients.find(fd) != clients.end())
-            client_read(fd, envp);
-          disconnect(fd);
+          if (clients.find(fd) != clients.end()) {
+            Result<Void> cr = client_read(fd, envp);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 16]
```diff
@@ -596,25 +273,37 @@ Result<Void> Server::start(char **envp) {
...
+            if (!cr.has_value())
+              return ERR(Void, cr.error());
+          }
+          Result<Void> d = disconnect(fd);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 17]
```diff
@@ -625,27 +314,44 @@ Result<Void> Server::start(char **envp) {
...
                   DefaultError::default_err_response(Response::GATEWAY_TIMEOUT);
-            else // res.error() == Errors::bad_gateway
+            else
               resp = DefaultError::default_err_response(Response::BAD_GATEWAY);
           } else {
             Result<std::string> output = cgi->poll();
+            std::map<FileDescriptor const *, ClientSession>::iterator jt =
+                clients.find(client_fd);
+            if (jt == clients.end()) {
+              Result<Void> rr = reap_cgi(cgi);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 18]
```diff
@@ -625,27 +314,44 @@ Result<Void> Server::start(char **envp) {
...
             oss << resp;
             if (!resp.keep_alive)
-              clients.at(client_fd).dropping = true;
-            clients.at(client_fd).out_buff = oss.str();
-            client_write(client_fd);
-            reap_cgi(cgi);
+              jt->second.dropping = true;
+            jt->second.out_buff = oss.str();
+            {
+              Result<Void> cw = client_write(client_fd);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server.cpp [hunk 19]
```diff
@@ -625,27 +314,44 @@ Result<Void> Server::start(char **envp) {
...
+              if (!cw.has_value())
+                return ERR(Void, cw.error());
+            }
+            {
+              Result<Void> rr = reap_cgi(cgi);
```
Comment (DavidLee18) [inline src/server/Server.cpp]: `TRY_`

File: src/server/Server_connections.cpp [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+      return ERR(Void, std::string("epoll add failed: ") + add_result.error());
+    }
+    const FileDescriptor *client_ptr = add_result.value();
+    if (listeners.find(server_fd) != listeners.end()) {
+      client.config = listeners.at(server_fd);
+    }
+    clients[client_ptr] = client;
+    std::cout << utils::info << "New client connected!" << std::endl;
+  }
+  return OK(Void, VOID);
```
Comment (DavidLee18) [inline src/server/Server_connections.cpp]: `OKV` 로 대체 가능

File: src/server/Server_connections.cpp [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+Result<Void> Server::disconnect(const FileDescriptor *client_fd) {
+  epoll.del_fd(client_fd);
+  clients.erase(client_fd);
+  return OK(Void, VOID);
```
Comment (DavidLee18) [inline src/server/Server_connections.cpp]: `OKV`

Comment (DavidLee18) [reply src/server/Server_connections.cpp]: 그리고 이런 경우는 `Result` 에러 처리를 하거나, 아니면 그냥 반환 타입을 `void` 로 하는 게 나아 보입니다.

File: src/server/Server_dispatch.cpp [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+
+  // Generate normal HTTP response
+  Response http = ResponseHandlers::http_response(client.req, &client,
+                                                  mime_type, &sessions, envp);
+  http.print_simple(std::cout);
+
+  delete client.req;
+  client.req = NULL;
+
+  Result<Void> qr = queue_response(client_fd, http);
```
Comment (DavidLee18) [inline src/server/Server_dispatch.cpp]: `TRY_`

File: src/server/Server_io.cpp [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+        if (client.dropping)
+          disconnect(client_fd);
+        return OK(Void, VOID);
+      }
+
+      dispatch_request(client_fd, client, envp);
+      // Client may have been disconnected in dispatch_request, check existence
+      if (clients.find(client_fd) == clients.end())
+        return OK(Void, VOID);
+      continue;
```
[the remark this answers is not in the document]
Comment (DavidLee18) [reply src/server/Server_io.cpp:96]: `TRY_` 사용 권장드립니다~

File: tests/webserv_cgi_framing_tests.zsh [hunk 24]
```diff
@@ -101,7 +101,7 @@ SOURCES=(
   "$PROJECT_ROOT/src/server/Server.cpp"
   "$PROJECT_ROOT/src/server/Session.cpp"
   "$PROJECT_ROOT/src/cgi_1_1.cpp"
-  "$PROJECT_ROOT/src/epoll_kqueue.cpp"
+  "$PROJECT_ROOT/src/EPoll.cpp"
   "$PROJECT_ROOT/src/framing_test_dummy.cpp"
```
[the remark this answers is not in the document]
Comment (DavidLee18) [reply tests/webserv_cgi_framing_tests.zsh:112]: 추후 수정 필요

File: sipyeon.test
Comment (DavidLee18) [inline sipyeon.test:1]: 이 파일은 뭔가요..?

**Review by DavidLee18 — CHANGES_REQUESTED** [review]
일단 구조가 상의 없이 너무 많이 바뀌어서 좀 당황스럽네요;; 그리고 매크로 활용해서 코드 짧게 만들 수 있는 부분은 그렇게 해 주세요. 그리고 저희 config 상 `HEAD` 메서드는 넣을 수 없습니다. 만약 config 부분도 동작을 수정하셨다면 원복해 주세요.

File: src/server/Server_io.cpp [hunk 25]
```diff
@@ -0,0 +1,269 @@
...
+      }
+
+      client.req = req_.value();
+      const Result<size_t> req_cl = client.req->get_content_length();
+      RouteRule rule;
+      Result<RouteRule> rule_res = client.config->find_route(
+          client.req->get_method(), client.req->get_path());
+      if (!rule_res.has_value())
+        return OK(Void, VOID);
+      rule = rule_res.value();
```
[the remark this answers is not in the document]
Comment (DavidLee18) [reply src/server/Server_io.cpp]: 이 부분도 확인해 주세요~

File: src/server/ResponseHandlers.cpp [hunk 26]
```diff
@@ -0,0 +1,557 @@
...
+    } else
+      return ResponseErrors::error_response(config, rule, Response::NOT_FOUND,
+                                            envp);
+  }
+
+  std::cout << utils::info << "matched rule path: '" << rule->path
+            << "' upload_dir = '" << rule->root.to_string() << "'" << std::endl;
+  if (rule->root.to_string().empty())
+    const_cast<RouteRule *>(rule)->root.to_string() = rule->root.to_string();
+  if (!rule->root.to_string().empty()) {
```
[the remark this answers is not in the document]
Comment (DavidLee18) [reply src/server/ResponseHandlers.cpp:156]: 확인 및 수정 필요

**Review by DavidLee18 — COMMENTED** [review]
**Overall Review: Excellent architectural refactor** (Sipyeon branch)

This PR significantly improves the codebase structure and maintainability. The split into `core/`, dedicated response handlers, and cleaner separation of concerns (connections, I/O, dispatch) is very welcome.

**Strengths:**
- Much better modularity (ResponseHandlers, ResponseUtils, ResponseErrors, etc.)
- Large file streaming support implemented
- Improved error handling with Result<> and TRY macros
- Cleaner autoindex and multipart parsing
- Good progress toward non-blocking I/O compliance

**I recommend APPROVE after addressing the points below.** (mostly small fixes + one rebase needed)

I'll add inline comments on specific files/lines below.

**Review by DavidLee18 — CHANGES_REQUESTED** [review]
**Overall Review: Request Changes**

Sir @sypyeon,

This is a very strong architectural overhaul of the webserv project (90% confidence it improves maintainability significantly). The split into `src/core/`, new `Server_connections.cpp` / `Server_io.cpp` / `Server_dispatch.cpp`, improved request parsing helpers, large-file streaming support, and polished Autoindex with delete buttons are excellent.

However, there are a few critical issues that need fixing before merge:

### Critical fixes required:
1. **Rebase onto latest `main`** - PR is currently behind (merge conflicts expected in `main.cpp` and Makefile).
2. **Compilation errors (`-Werror`)**:
   - `main.cpp`: unused variable `temp`.
   - `src/server/ResponseHandlers.cpp` (around line 158): temporary-assignment bug on `rule->root.to_string()` in `post_method_response`.
3. Several `Result<>` returns are ignored in critical paths (e.g. `queue_response`, `client_read` fallbacks in dispatch logic). Please propagate errors properly.

### High priority improvements:
- `If-Modified-Since` header comparison should use numeric mtime instead of string comparison.
- Review edge cases for `.cgi` extension handling and `client_max_body_size` parsing.
- Update outdated test files (e.g. `webserv_cgi_framing_tests.zsh`) to match new directory layout.

### Positive highlights:
- Large-file streaming implementation via `ClientSession` (`out_file_path` / `out_file_offset` / `streaming_file`) is exactly what the non-blocking requirements demand.
- Much cleaner responsibility separation with the new `Response*` classes and `MultipartParser`.
- Autoindex HTML is now properly escaped and includes JS delete functionality - very nice touch.

Once these are addressed, this PR should be in excellent shape for evaluation.

Happy to do a more detailed line-by-line review on specific modules if needed.

**Review by DavidLee18 — COMMENTED** [review]
Test - confirming GitHub agent capability

File: src/server/ResponseHandlers.cpp (hunk 5 above)
Comment (sypyeon) [reply src/server/ResponseHandlers.cpp:26]: no default constructor available

File: tests/webserv_cgi_framing_tests.zsh (hunk 24 above)
[the remark this answers is not in the document]
Comment (DavidLee18) [reply tests/webserv_cgi_framing_tests.zsh:112]: will do later, not critical

File: src/server/Server.cpp [hunk 27]
```diff
@@ -467,59 +88,74 @@ Result<Void> Server::start(char **envp) {
...
+      if (jt == clients.end())
+        continue;
+      std::ostringstream oss;
+      oss << resp;
+      if (!resp.keep_alive)
+        jt->second.dropping = true;
+      jt->second.out_buff = oss.str();
+      TRY_(Void, Void, client_write(jt->first))
+      if (clients.find(fd) != clients.end() && jt->second.dropping)
+        TRY_(Void, Void, disconnect(fd))
```
[the remark this answers is not in the document]
Comment (DavidLee18) [reply src/server/Server.cpp:144]: will do, but on my branch or after this merges.
